### Case study  - Uber Data Analysis

The data of a driver’s uber trips are available for year 2016.
Your manager wants you to explore this data to give him some useful insights about the trip behaviour of a Uber driver.

#### Dataset - 
The dataset contains Start Date, End Date, Start Location, End Location, Miles Driven and Purpose of drive (Business, Personal, Meals, Errands, Meetings, Customer Support etc.)


#### Steps - 

1.Import the libraries

2.Get the data and observe it

3.Check missing values, either remove it or fill it.

4.Get summary of data using python function.

5.Explore the data parameter wise

Here we have information of destination(start and stop), time(start and stop), category and purpose of trip, miles covered.


In [39]:
# ----------------------
# Concepts To cover 
# ----------------------
# 1. Data profiling
# 2. group by function
# 3. Apply function 
# 4. DateTime operations 

In [40]:
# Import the libraries 
import numpy as np
import pandas as pd

In [41]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [42]:
a= 10
b=10
a
b

10

10

In [43]:
# Read the Data 


# df = pd.read_csv('../data/uberdrive.csv', skiprows = 100)  # To skip first n nows 
# df = pd.read_csv('../data/uberdrive.csv', nrows = 100)  # Load only the first n rows
# df = pd.read_csv('../data/uberdrive.csv', usecols = ['START_DATE*', 'END_DATE*'])  # Load only selected columns

df = pd.read_csv('uberdrive.csv')

In [44]:
df.head(8) # View first n rows of data 

START_DATE*       END_DATE* CATEGORY*           START*            STOP*  \
0  1/1/2016 21:11  1/1/2016 21:17  Business      Fort Pierce      Fort Pierce   
1   1/2/2016 1:25   1/2/2016 1:37  Business      Fort Pierce      Fort Pierce   
2  1/2/2016 20:25  1/2/2016 20:38  Business      Fort Pierce      Fort Pierce   
3  1/5/2016 17:31  1/5/2016 17:45  Business      Fort Pierce      Fort Pierce   
4  1/6/2016 14:42  1/6/2016 15:49  Business      Fort Pierce  West Palm Beach   
5  1/6/2016 17:15  1/6/2016 17:19  Business  West Palm Beach  West Palm Beach   
6  1/6/2016 17:30  1/6/2016 17:35  Business  West Palm Beach       Palm Beach   
7  1/7/2016 13:27  1/7/2016 13:33  Business             Cary             Cary   

   MILES*         PURPOSE*  
0     5.1   Meal/Entertain  
1     5.0              NaN  
2     4.8  Errand/Supplies  
3     4.7          Meeting  
4    63.7   Customer Visit  
5     4.3   Meal/Entertain  
6     7.1          Meeting  
7     0.8          Meeting

In [45]:
#View the last 5 rows of data
df.tail()

START_DATE*         END_DATE* CATEGORY*            START*  \
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                 STOP*   MILES*        PURPOSE*  
1151  Unknown Location      3.9  Temporary Site  
1152  Unknown Location     16.2         Meeting  
1153           Gampaha      6.4  Temporary Site  
1154         Ilukwatta     48.2  Temporary Site  
1155               NaN  12204.7             NaN

In [46]:
# understand shape and size of data 
print(df.shape)
print (df.size)

(1156, 7)
8092


In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1156 entries, 0 to 1155
Data columns (total 7 columns):
START_DATE*    1156 non-null object
END_DATE*      1155 non-null object
CATEGORY*      1155 non-null object
START*         1155 non-null object
STOP*          1155 non-null object
MILES*         1156 non-null float64
PURPOSE*       653 non-null object
dtypes: float64(1), object(6)
memory usage: 63.3+ KB


In [48]:
# Same as above, gives non-null number of records
df.count()

START_DATE*    1156
END_DATE*      1155
CATEGORY*      1155
START*         1155
STOP*          1155
MILES*         1156
PURPOSE*        653
dtype: int64

In [49]:
df.dtypes

START_DATE*     object
END_DATE*       object
CATEGORY*       object
START*          object
STOP*           object
MILES*         float64
PURPOSE*        object
dtype: object

1. PURPOSE column has lots of missing values  
2. 1155 or 1156 records ??

In [50]:
df.isnull().sum()

START_DATE*      0
END_DATE*        1
CATEGORY*        1
START*           1
STOP*            1
MILES*           0
PURPOSE*       503
dtype: int64

In [51]:
#Is any of the values in the df null ? 
df.isnull().any() #per column

START_DATE*    False
END_DATE*       True
CATEGORY*       True
START*          True
STOP*           True
MILES*         False
PURPOSE*        True
dtype: bool

In [52]:
#Is any of the values in the df null ?  ( # Useful in writing validation scripts on large number of files )
df.isnull().any().any()

True

In [53]:
#Is any of the values in columns of the df null ? ( # Useful in writing validation scripts on large number of files )
df.isnull().any() 

START_DATE*    False
END_DATE*       True
CATEGORY*       True
START*          True
STOP*           True
MILES*         False
PURPOSE*        True
dtype: bool

In [54]:
null_cols = df.columns[df.isnull().any()]
null_cols = list(null_cols)
#Get the columns into a list and do use it to do some operations 

In [55]:
df.isnull().sum()  #Shows  the column wise values of missing data

START_DATE*      0
END_DATE*        1
CATEGORY*        1
START*           1
STOP*            1
MILES*           0
PURPOSE*       503
dtype: int64

Show the records with missing values for column= PURPOSE 

In [56]:
df.count()

START_DATE*    1156
END_DATE*      1155
CATEGORY*      1155
START*         1155
STOP*          1155
MILES*         1156
PURPOSE*        653
dtype: int64

In [57]:
df.iloc[[0,2,3,4],[1,2]]

END_DATE* CATEGORY*
0  1/1/2016 21:17  Business
2  1/2/2016 20:38  Business
3  1/5/2016 17:45  Business
4  1/6/2016 15:49  Business

In [58]:
df.iloc[100:200][["START_DATE*","END_DATE*"]]

START_DATE*        END_DATE*
100  2/12/2016 15:33  2/12/2016 16:06
101  2/13/2016 14:21  2/13/2016 14:41
102  2/13/2016 23:45   2/14/2016 0:01
103   2/14/2016 0:50   2/14/2016 1:00
104  2/14/2016 14:07  2/14/2016 14:40
..               ...              ...
195   3/5/2016 14:08   3/5/2016 14:18
196   3/5/2016 14:39   3/5/2016 15:01
197   3/5/2016 16:52   3/5/2016 17:13
198   3/5/2016 17:23   3/5/2016 17:34
199    3/7/2016 9:10    3/7/2016 9:20

[100 rows x 2 columns]

In [59]:
df[(df['PURPOSE*'].isnull())]
# you can as well do a head() function on this to just view the first n rows

START_DATE*         END_DATE* CATEGORY*            START*  \
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
32      1/19/2016 9:09    1/19/2016 9:23  Business       Whitebridge   
85      2/9/2016 10:54    2/9/2016 11:07  Personal       Whitebridge   
86      2/9/2016 11:43    2/9/2016 11:50  Personal        Northwoods   
87      2/9/2016 13:36    2/9/2016 13:52  Personal        Tanglewood   
...                ...               ...       ...               ...   
1066  12/19/2016 14:37  12/19/2016 14:50  Business  Unknown Location   
1069  12/19/2016 19:05  12/19/2016 19:17  Business         Islamabad   
1071   12/20/2016 8:49   12/20/2016 9:24  Business  Unknown Location   
1143  12/29/2016 20:53  12/29/2016 21:42  Business           Kar?chi   
1155            Totals               NaN       NaN               NaN   

                    STOP*   MILES* PURPOSE*  
1             Fort Pierce      5.0      NaN  
32    Lake Wellingborough      7.2      NaN  
85             Northwoods      5.3      NaN  
86             Tanglewood      3.0      NaN  
87                Preston      5.1      NaN  
...                   ...      ...      ...  
1066     Unknown Location      5.4      NaN  
1069     Unknown Location      2.2      NaN  
1071           Rawalpindi     12.0      NaN  
1143     Unknown Location      6.4      NaN  
1155                  NaN  12204.7      NaN  

[503 rows x 7 columns]

How many records are non-null  / have values ( in a particular column )

In [60]:
df['PURPOSE*'].count()

653

### Renaming columns

In [ ]:
df.str

In [61]:
# Rename the columns to remove the * from the names
# Why ? we can use df.START_DATE same way as  df['START_DATE*']
# col_names = ['START_DATE','END_DATE','CAT', 'START' , 'STOP' , 'MILES','PURPOSE']
# df.columns = col_names


#or 
# Replace the * character from all the  columns .
# ( in case you have many columns and cant manually write the column names ) ( Thanks for pointing this out.)
df.columns = df.columns.str.replace("*", "")

# You can also rename the specific column names 
df.rename( columns = {'CAT':'CATEGORY'})
df.head()

START_DATE          END_DATE  CATEGORY             START  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
...                ...               ...       ...               ...   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                  STOP    MILES          PURPOSE  
0          Fort Pierce      5.1   Meal/Entertain  
1          Fort Pierce      5.0              NaN  
2          Fort Pierce      4.8  Errand/Supplies  
3          Fort Pierce      4.7          Meeting  
4      West Palm Beach     63.7   Customer Visit  
...                ...      ...              ...  
1151  Unknown Location      3.9   Temporary Site  
1152  Unknown Location     16.2          Meeting  
1153           Gampaha      6.4   Temporary Site  
1154         Ilukwatta     48.2   Temporary Site  
1155               NaN  12204.7              NaN  

[1156 rows x 7 columns]

START_DATE        END_DATE  CATEGORY        START             STOP  \
0  1/1/2016 21:11  1/1/2016 21:17  Business  Fort Pierce      Fort Pierce   
1   1/2/2016 1:25   1/2/2016 1:37  Business  Fort Pierce      Fort Pierce   
2  1/2/2016 20:25  1/2/2016 20:38  Business  Fort Pierce      Fort Pierce   
3  1/5/2016 17:31  1/5/2016 17:45  Business  Fort Pierce      Fort Pierce   
4  1/6/2016 14:42  1/6/2016 15:49  Business  Fort Pierce  West Palm Beach   

   MILES          PURPOSE  
0    5.1   Meal/Entertain  
1    5.0              NaN  
2    4.8  Errand/Supplies  
3    4.7          Meeting  
4   63.7   Customer Visit

In [62]:
df.head()

START_DATE        END_DATE  CATEGORY        START             STOP  \
0  1/1/2016 21:11  1/1/2016 21:17  Business  Fort Pierce      Fort Pierce   
1   1/2/2016 1:25   1/2/2016 1:37  Business  Fort Pierce      Fort Pierce   
2  1/2/2016 20:25  1/2/2016 20:38  Business  Fort Pierce      Fort Pierce   
3  1/5/2016 17:31  1/5/2016 17:45  Business  Fort Pierce      Fort Pierce   
4  1/6/2016 14:42  1/6/2016 15:49  Business  Fort Pierce  West Palm Beach   

   MILES          PURPOSE  
0    5.1   Meal/Entertain  
1    5.0              NaN  
2    4.8  Errand/Supplies  
3    4.7          Meeting  
4   63.7   Customer Visit

### Filtering dataframes -1 

In [63]:
# shows the entries where PURPOSE is null
df[df.PURPOSE.isnull()]

# inverting the selection ( not null ) ( works for booleans cases)
df[~df.PURPOSE.isnull()]

START_DATE          END_DATE  CATEGORY             START  \
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
32      1/19/2016 9:09    1/19/2016 9:23  Business       Whitebridge   
85      2/9/2016 10:54    2/9/2016 11:07  Personal       Whitebridge   
86      2/9/2016 11:43    2/9/2016 11:50  Personal        Northwoods   
87      2/9/2016 13:36    2/9/2016 13:52  Personal        Tanglewood   
...                ...               ...       ...               ...   
1066  12/19/2016 14:37  12/19/2016 14:50  Business  Unknown Location   
1069  12/19/2016 19:05  12/19/2016 19:17  Business         Islamabad   
1071   12/20/2016 8:49   12/20/2016 9:24  Business  Unknown Location   
1143  12/29/2016 20:53  12/29/2016 21:42  Business           Kar?chi   
1155            Totals               NaN       NaN               NaN   

                     STOP    MILES PURPOSE  
1             Fort Pierce      5.0     NaN  
32    Lake Wellingborough      7.2     NaN  
85             Northwoods      5.3     NaN  
86             Tanglewood      3.0     NaN  
87                Preston      5.1     NaN  
...                   ...      ...     ...  
1066     Unknown Location      5.4     NaN  
1069     Unknown Location      2.2     NaN  
1071           Rawalpindi     12.0     NaN  
1143     Unknown Location      6.4     NaN  
1155                  NaN  12204.7     NaN  

[503 rows x 7 columns]

START_DATE          END_DATE  CATEGORY             START  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
5       1/6/2016 17:15    1/6/2016 17:19  Business   West Palm Beach   
...                ...               ...       ...               ...   
1150   12/31/2016 1:07   12/31/2016 1:14  Business           Kar?chi   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   

                  STOP  MILES          PURPOSE  
0          Fort Pierce    5.1   Meal/Entertain  
2          Fort Pierce    4.8  Errand/Supplies  
3          Fort Pierce    4.7          Meeting  
4      West Palm Beach   63.7   Customer Visit  
5      West Palm Beach    4.3   Meal/Entertain  
...                ...    ...              ...  
1150           Kar?chi    0.7          Meeting  
1151  Unknown Location    3.9   Temporary Site  
1152  Unknown Location   16.2          Meeting  
1153           Gampaha    6.4   Temporary Site  
1154         Ilukwatta   48.2   Temporary Site  

[653 rows x 7 columns]

### Filtering dataframe - 2 

In [64]:
#1. Conditions within dataframe 
df[df['MILES'] > 30]

#2. SQL-like query
df.query('MILES > 30')

# inverting the selection 
df[df.MILES == 30 ]
df[df.MILES != 30 ]

START_DATE          END_DATE  CATEGORY             START  \
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
36     1/20/2016 13:25   1/20/2016 14:19  Business           Raleigh   
108     2/16/2016 3:21    2/16/2016 4:13  Business        Katunayaka   
232    3/17/2016 12:52   3/17/2016 15:11  Business            Austin   
233    3/17/2016 15:16   3/17/2016 15:58  Business              Katy   
245    3/19/2016 14:01   3/19/2016 14:57  Business        Sugar Land   
246    3/19/2016 15:34   3/19/2016 16:38  Business           Houston   
251    3/19/2016 19:33   3/19/2016 20:39  Business         Galveston   
268    3/25/2016 13:24   3/25/2016 16:22  Business              Cary   
269    3/25/2016 16:52   3/25/2016 22:22  Business             Latta   
270    3/25/2016 22:54    3/26/2016 1:39  Business      Jacksonville   
295     4/2/2016 12:21    4/2/2016 14:47  Business         Kissimmee   
296     4/2/2016 16:57    4/2/2016 18:09  Business     Daytona Beach   
297     4/2/2016 19:38    4/2/2016 22:36  Business      Jacksonville   
298     4/2/2016 23:11     4/3/2016 1:34  Business         Ridgeland   
299      4/3/2016 2:00     4/3/2016 4:16  Business          Florence   
335    4/28/2016 13:30   4/28/2016 13:49  Business           Raleigh   
478     6/27/2016 7:37    6/27/2016 8:48  Business       New Orleans   
546    7/14/2016 16:39   7/14/2016 20:05  Business       Morrisville   
554    7/16/2016 15:29   7/16/2016 16:57  Personal           Newland   
559    7/17/2016 12:20   7/17/2016 15:25  Personal             Boone   
651    8/11/2016 18:37   8/11/2016 19:29  Business              Cary   
652    8/11/2016 20:59   8/11/2016 21:44  Business       Wake Forest   
707    8/24/2016 13:01   8/24/2016 15:25  Business  Unknown Location   
708    8/25/2016 15:17   8/25/2016 16:22  Business  Unknown Location   
710    8/25/2016 17:19   8/25/2016 19:20  Business  Unknown Location   
726    8/27/2016 14:01   8/27/2016 15:44  Business            Lahore   
727    8/27/2016 16:15   8/27/2016 19:13  Business  Unknown Location   
751     9/6/2016 17:49    9/6/2016 17:49  Business  Unknown Location   
776    9/27/2016 21:01    9/28/2016 2:37  Business  Unknown Location   
779    9/30/2016 17:39   9/30/2016 20:20  Business         Islamabad   
788    10/6/2016 17:23   10/6/2016 17:40  Business        R?walpindi   
791    10/7/2016 10:56   10/7/2016 10:59  Business  Unknown Location   
841   10/20/2016 12:19  10/20/2016 13:17  Business          Berkeley   
842   10/20/2016 20:44  10/20/2016 21:37  Business          San Jose   
869   10/28/2016 15:53  10/28/2016 17:59  Business              Cary   
870   10/28/2016 18:13  10/28/2016 20:07  Business     Winston Salem   
871   10/28/2016 20:13  10/28/2016 22:00  Business         Asheville   
872   10/29/2016 15:22  10/29/2016 17:05  Business            Topton   
873   10/29/2016 17:13  10/29/2016 19:19  Business        Hayesville   
880   10/30/2016 13:24  10/30/2016 14:37  Business       Bryson City   
881   10/30/2016 15:22  10/30/2016 18:23  Business         Asheville   
882   10/30/2016 18:26  10/30/2016 19:39  Business            Mebane   
913    11/6/2016 16:27   11/6/2016 17:28  Business       Santa Clara   
936    11/11/2016 9:35  11/11/2016 10:23  Business          Berkeley   
939   11/11/2016 14:39  11/11/2016 15:46  Business        Menlo Park   
960   11/14/2016 11:24  11/14/2016 12:13  Business          Berkeley   
961   11/14/2016 13:40  11/14/2016 14:33  Business     Mountain View   
979   11/20/2016 10:27  11/20/2016 11:32  Business              Cary   
1088  12/21/2016 20:56  12/21/2016 23:42  Business        Rawalpindi   
1089  12/22/2016 15:40  12/22/2016 16:38  Business  Unknown Location   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                  STOP    MILES          PURPOSE  
4      West Palm Beach     63.7   Customer Visit  
36                Car

START_DATE          END_DATE  CATEGORY             START  \
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
36     1/20/2016 13:25   1/20/2016 14:19  Business           Raleigh   
108     2/16/2016 3:21    2/16/2016 4:13  Business        Katunayaka   
232    3/17/2016 12:52   3/17/2016 15:11  Business            Austin   
233    3/17/2016 15:16   3/17/2016 15:58  Business              Katy   
245    3/19/2016 14:01   3/19/2016 14:57  Business        Sugar Land   
246    3/19/2016 15:34   3/19/2016 16:38  Business           Houston   
251    3/19/2016 19:33   3/19/2016 20:39  Business         Galveston   
268    3/25/2016 13:24   3/25/2016 16:22  Business              Cary   
269    3/25/2016 16:52   3/25/2016 22:22  Business             Latta   
270    3/25/2016 22:54    3/26/2016 1:39  Business      Jacksonville   
295     4/2/2016 12:21    4/2/2016 14:47  Business         Kissimmee   
296     4/2/2016 16:57    4/2/2016 18:09  Business     Daytona Beach   
297     4/2/2016 19:38    4/2/2016 22:36  Business      Jacksonville   
298     4/2/2016 23:11     4/3/2016 1:34  Business         Ridgeland   
299      4/3/2016 2:00     4/3/2016 4:16  Business          Florence   
335    4/28/2016 13:30   4/28/2016 13:49  Business           Raleigh   
478     6/27/2016 7:37    6/27/2016 8:48  Business       New Orleans   
546    7/14/2016 16:39   7/14/2016 20:05  Business       Morrisville   
554    7/16/2016 15:29   7/16/2016 16:57  Personal           Newland   
559    7/17/2016 12:20   7/17/2016 15:25  Personal             Boone   
651    8/11/2016 18:37   8/11/2016 19:29  Business              Cary   
652    8/11/2016 20:59   8/11/2016 21:44  Business       Wake Forest   
707    8/24/2016 13:01   8/24/2016 15:25  Business  Unknown Location   
708    8/25/2016 15:17   8/25/2016 16:22  Business  Unknown Location   
710    8/25/2016 17:19   8/25/2016 19:20  Business  Unknown Location   
726    8/27/2016 14:01   8/27/2016 15:44  Business            Lahore   
727    8/27/2016 16:15   8/27/2016 19:13  Business  Unknown Location   
751     9/6/2016 17:49    9/6/2016 17:49  Business  Unknown Location   
776    9/27/2016 21:01    9/28/2016 2:37  Business  Unknown Location   
779    9/30/2016 17:39   9/30/2016 20:20  Business         Islamabad   
788    10/6/2016 17:23   10/6/2016 17:40  Business        R?walpindi   
791    10/7/2016 10:56   10/7/2016 10:59  Business  Unknown Location   
841   10/20/2016 12:19  10/20/2016 13:17  Business          Berkeley   
842   10/20/2016 20:44  10/20/2016 21:37  Business          San Jose   
869   10/28/2016 15:53  10/28/2016 17:59  Business              Cary   
870   10/28/2016 18:13  10/28/2016 20:07  Business     Winston Salem   
871   10/28/2016 20:13  10/28/2016 22:00  Business         Asheville   
872   10/29/2016 15:22  10/29/2016 17:05  Business            Topton   
873   10/29/2016 17:13  10/29/2016 19:19  Business        Hayesville   
880   10/30/2016 13:24  10/30/2016 14:37  Business       Bryson City   
881   10/30/2016 15:22  10/30/2016 18:23  Business         Asheville   
882   10/30/2016 18:26  10/30/2016 19:39  Business            Mebane   
913    11/6/2016 16:27   11/6/2016 17:28  Business       Santa Clara   
936    11/11/2016 9:35  11/11/2016 10:23  Business          Berkeley   
939   11/11/2016 14:39  11/11/2016 15:46  Business        Menlo Park   
960   11/14/2016 11:24  11/14/2016 12:13  Business          Berkeley   
961   11/14/2016 13:40  11/14/2016 14:33  Business     Mountain View   
979   11/20/2016 10:27  11/20/2016 11:32  Business              Cary   
1088  12/21/2016 20:56  12/21/2016 23:42  Business        Rawalpindi   
1089  12/22/2016 15:40  12/22/2016 16:38  Business  Unknown Location   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                  STOP    MILES          PURPOSE  
4      West Palm Beach     63.7   Customer Visit  
36                Car

START_DATE         END_DATE  CATEGORY       START      STOP  MILES  \
486  6/27/2016 12:22  6/27/2016 13:02  Business  Mandeville  Metairie   30.0   

    PURPOSE  
486     NaN

START_DATE          END_DATE  CATEGORY             START  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
...                ...               ...       ...               ...   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                  STOP    MILES          PURPOSE  
0          Fort Pierce      5.1   Meal/Entertain  
1          Fort Pierce      5.0              NaN  
2          Fort Pierce      4.8  Errand/Supplies  
3          Fort Pierce      4.7          Meeting  
4      West Palm Beach     63.7   Customer Visit  
...                ...      ...              ...  
1151  Unknown Location      3.9   Temporary Site  
1152  Unknown Location     16.2          Meeting  
1153           Gampaha      6.4   Temporary Site  
1154         Ilukwatta     48.2   Temporary Site  
1155               NaN  12204.7              NaN  

[1155 rows x 7 columns]

Explore the details from the MILES column


In [65]:
#pd.Index([1, 2, 3])

In [ ]:
df.MILES.max()
df.MILES.sort_values(ascending = False).head(10) # Show the top 10 rides (*in terms of distance driven)
df[df.MILES ==df.MILES.max()]  # Show the row that has the max miles 
df.sort_values(by = 'MILES' , ascending = False).head(10) # Shows the top 10 rows of MILES ( decreasing value )

Remove the last row which has some anomalous number in MILES 

In [ ]:
len(df)
df = df [df.MILES < 1000] # 
# Or use other filters to remove this row

In [ ]:
len(df)

In [ ]:
df.dtypes

In [66]:
#Get a summary of the numerical columns in the data
df.describe()

MILES
count   1156.000000
mean      21.115398
std      359.299007
min        0.500000
25%        2.900000
50%        6.000000
75%       10.400000
max    12204.700000

The miles driven ranges from 0.5 miles to 12204 miles with an average of 21 miles

Max miles - looks suspect

#### Dropping rows  which have null values

In [ ]:
# Get the initial data with dropping the NA values
df_dropped = df.dropna()  # Dont do this on the original dataframe, you will lose good rows with values , make a copy

#Get the shape of the dataframe after removing the null values
df_dropped.shape

The filtered dataset with no nulls ( in PURPOSE column )  contains 653 rows of non-null values

### PANDAS PROFILING
Use some packages that will auto profile your data ( than having to manually type in all the commands )

You can find the documentation here ( https://pandas-profiling.github.io/pandas-profiling/docs/ )

Github page here - https://github.com/pandas-profiling/pandas-profiling

In [ ]:
import pandas_profiling 

#pandas_profiling.ProfileReport(df)  # Of you just want to show here 
#Or if you want to 
profile= pandas_profiling.ProfileReport(df)
profile.to_file('./output.html')   # output into an html file 

### Lets explore the data parameter wise - 

1.Destination - (starting and stopping)

2.Time - (hour of the day, day of week, month of year)

3.Categories

4.Purpose 

5.Grouping two parameters to get more insights


## 1. Understanding  the start and stop points 

In [ ]:
# Get the unique starting point, unique destination
# names of unique start points
print(df['START'].unique())

In [ ]:
len(df['START'].unique()) #count of unique start points using  len()

In [ ]:
print(df['START'].nunique()) # or use can use the nunique function

In [ ]:
# Get the starting destination, unique destination
print(df['STOP'].unique()) #names of unique start points

In [ ]:
print(len(df['STOP'].unique())) #count of unique start points

In [ ]:
# To take it one level up in the ease of comparing between two series / lists , use sets 
# It has 
# Find details comparing (between) unique start and stop locations 
# Details about start and stop locations 

s_start  = set(df['START']) #names of unique start points
s_stop =  set(df['STOP']) #names of unique start points

print(len(s_start))
print(len(s_stop))

In [ ]:
#Example on how to use sets to do intersection , union and difference

l1 = [1,2,3,4,4]
l2 = [3,4,5,6]

set(l2) - set(l1) # difference  ( whats in l2 and not in l1 )
set(l1) - set(l2) # difference ( whats in l1 and not in l2 )
set(l2) | set(l1) # union 
set(l2) & set(l1) # intersection 


Stations which are appeared in both start and stop locations 

In [ ]:
s_stop & s_start 

In [ ]:
#Identify popular start points - top 10
df['START'].value_counts().head(10)

In [ ]:
#Identify popular stop destinations - top 10
df['STOP'].value_counts().head(10)

In [ ]:
# Are there cases where the start and the stop location are the same  ? 
df[df['START'] == df['STOP']]

In [ ]:
# Favorite starting point wrt the total miles covered 
df.groupby('START')['MILES'].sum().sort_values(ascending = False ).head(10)

In [ ]:
ser = pd.Series(np.arange(10,20,2))

In [ ]:
ser

In [ ]:
ser = (ser.reset_index())

In [ ]:
ser

In [ ]:
type(df.groupby('START')['MILES'].sum().sort_values(ascending = False ))

#### Find out most farthest start and stop pair - top10 ( aggregation ) ( BY TOTAL miles COVERED EVER ! )

In [ ]:
#Doing a quick plot 
k3 = df.groupby('START')['MILES'].sum().sort_values(ascending=False).head(10) 
k3= k3.reset_index() # flatten the dataframe 
k3.columns = ['START' ,'sum_of_miles']
%matplotlib inline 
import seaborn as sns
import matplotlib.pyplot as plt
sns.barplot(data= k3 , x= 'START' , y ='sum_of_miles')
plt.xticks(rotation=90);

In [ ]:
sns.barplot(data= k3 , y= 'START' , x ='sum_of_miles',orient="h")
plt.xticks(rotation=90);

In [ ]:
df.groupby(['START','STOP'])['MILES'].sum().sort_values(ascending=False).head()

In [ ]:
#Dropping Unknown Location Value  - Save into anothe dataframe ( you dont want to overwrite the original df)
df2 = df[df['START']!= 'Unknown Location']
df2 = df2[df2['STOP']!= 'Unknown Location']

df2.groupby(['START','STOP'])['MILES'].sum().sort_values(ascending=False).head(10)

The most popular start and stop pair - ( BY COUNT of travels! )

In [ ]:
df2.groupby(['START','STOP'])["MILES"].count().sort_values(ascending=False)

In [ ]:
#Sum of miles
df2.groupby(['START','STOP'])['MILES'].sum().sort_values(ascending=False).head(10)

In [ ]:
df2.groupby(['START','STOP'])['MILES']['MILES'].size().sort_values(ascending=False).head(10)

**The most popular start to destination pair is Morrisville-Cary**

## 2. Manipulating date & time objects

In [ ]:
df.head()

We will be using the datetime module (  https://docs.python.org/3/library/datetime.html  )

Search for "strftime() and strptime() Format Codes" in the documentation page

In [ ]:
df.dtypes

In [ ]:
# Create columns by converting the start and end date into a datatime format
# You can also over write the same column - but for the sake of understanding the difference in formats , we create new columns
from datetime import datetime
df['start_dt'] = df['START_DATE'].apply(lambda x : datetime.strptime(x, '%m/%d/%Y %H:%M'))
df['end_dt'] = df['END_DATE'].apply(lambda x : datetime.strptime(x, '%m/%d/%Y %H:%M'))

In [ ]:
df.head()
df.dtypes  # See how the dtype is different ?

In [ ]:
#Create more columns by using the inbuilt functionalities of datatime module

df['start_day'] = df['start_dt'].dt.day
df['start_hour'] = df['start_dt'].dt.hour
df['start_month'] = df['start_dt'].dt.month
df['d_of_wk'] = df['start_dt'].dt.dayofweek   # Days encoded as 0-6  ( monday =0, Tue =1 .... )


In [ ]:
# You can convert the numeric encoding of weekdays into short form by manually writing the mapping 
# days = {0:'Mon',1:'Tue',2:'Wed',3:'Thur',4:'Fri',5:'Sat',6:'Sun'}
# df['day_of_week'] = df['day_of_week'].apply(lambda x: days[x])

#or use the builtin functions 

df['weekday'] = df['start_dt'].apply(lambda x : datetime.strftime(x,'%a'))  # ( or directly convert into the short form)

In [ ]:
# Similarly you can use the calendar library to get the month abbreviation 

#import calendar
#df['start_month_cal'] = df['start_month'].apply(lambda x: calendar.month_abbr[x])

# or 
# use the built in functions in datatime module
df['cal_month'] =  df['start_dt'].apply(lambda x : datetime.strftime(x,'%b'))

In [ ]:
df.head()

In [ ]:
# Which month did he get most drives  ? 
df.groupby(['start_month']).size() #['start_day']

In [ ]:
df.groupby('cal_month').sum()['MILES'].sort_values(ascending = False)

In [ ]:
# Getting the average distance covered each month
df.groupby('cal_month').mean()['MILES'].sort_values(ascending = False)

In [ ]:
# Write your inferences here based on the last 3 outputs 

In [ ]:
#Which day did he get most drives  ? 
df.groupby(['weekday']).size()  

**When does he usually start the trip ?**

In [ ]:
df.groupby('start_hour').size()

This looks like an interesting data 

Does he have a prefered time of start ?

In [ ]:
df['stop_hour'] = df['end_dt'].dt.hour

In [ ]:
df_hrs = df.groupby('stop_hour').size()
df_hrs = df_hrs.reset_index()
df_hrs.columns = ['start_hour' ,'count']
sns.barplot(data= df_hrs , x ='start_hour' , y = 'count')

Looks like he mostly starts the trip around 9-10 and the peak hours seem to be between 12- 5 PM 

### Duration of the trips 

In [ ]:
df['diff'] = (df['end_dt'] - df['start_dt'])

In [ ]:
df.dtypes

This creates a timedelta datatype

### How long did the trips last 

In [ ]:
df['diff_hr'] = df['diff'].astype('timedelta64[h]')

Find the date time units in  https://docs.scipy.org/doc/numpy/reference/arrays.datetime.html

search for 'Datetime Units'

In [ ]:
df['diff'].describe()

In [ ]:
# View in terms of minutes 
df['diff_mins'] = df['diff'].astype('timedelta64[m]')

In [ ]:
df['diff_mins'].describe()

There seems to be somethin strange with the minumum time ( in minutes ) - it is 0 

In [ ]:
len(df[df['START_DATE'] == df['END_DATE']])

df[df['START_DATE'] == df['END_DATE']]

There are some values where the start and end dates are the same - remove them ?

In [ ]:
len(df2)

### Speed 

In [ ]:
# calculate trip speed for each trip
df['Duration_hours'] = df['diff_mins'] / 60
df['Speed'] = df['MILES'] / df['Duration_hours']
df['Speed'].describe()

Remove the ones with unknown location

In [ ]:
# df2 = df[df.START != 'Unknown Location']
# df2 = df2[df2.STOP != 'Unknown Location']

#or 
#you can filter using one single conditional statement
df2 = df[(df.START != 'Unknown Location') & ( df.STOP != 'Unknown Location')]

In [ ]:
# Remove the ones that have the same start and end time  ( and redo the analysis )
df2 = df[df.start_dt != df.end_dt]


# we still see some really fast driving !! 
df2.sort_values(by ='diff_mins', ascending = True).head(5)

# To do 
# Take a cut off for the time taken to drive to  do the speed analysis again 
# something like ( df.diff_mins > 5 ) # analyse only the trips that lasted atleast 5 mins 

## 3. Category & Purpose

In [ ]:
df['CAT'].value_counts()

**Most trips are in the business category**

In [ ]:
#Purpose
df['PURPOSE'].value_counts()

**Most trips are for meetings**

In [ ]:
#Average distance traveled for each activity
df.groupby('PURPOSE').mean()['MILES'].sort_values(ascending = False)

Now lets try to answer some questions from this data.

Question1: How many miles was earned per category and purpose ?

Question2: What is percentage of business miles vs personal?

Question3: How much time was spend for drives per category and purpose? 


In [ ]:
#Question1: How many miles was earned per category and purpose ?
df.groupby('PURPOSE').sum()['MILES'].sort_values(ascending = False)

In [ ]:
#Question1: How many miles was earned per category and purpose ?
df.groupby('CAT').sum()['MILES'].sort_values(ascending = False)

In [ ]:
#Question2: What is percentage of business miles vs personal?
df1 = df.groupby('CAT').agg({'MILES':'sum'})
df1
df1.apply(lambda x: x/x.sum()*100).rename(columns = {'MILES':'% of Miles'})

In [ ]:
#Question3: How much time was spend for drives per category and purpose?
dfk= df.groupby('CAT').sum()['diff_mins'].sort_values(ascending = False)
dfk.apply(lambda x: x/60)

In [ ]:
#Question3: How much time was spend for drives per category and purpose? 
df.groupby('PURPOSE').sum()['diff_mins'].sort_values(ascending = False)